untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,461 @@
+# Untyped Lambda Calculus
+
+When talking about Haskell, the term "lambda calculus" often crops up.
+It is a theoretical framework used to define the meaning of "computation" in many functional languages.
+Understanding lambda calculus can be very helpful when talking about Haskell programs.
+It is also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus, _typed_, and _untyped_.
+The _untyped_ kind, discussed in this article, expresses an unrestricted computation.
+However, as a system of logic, it is self-contradictory.
+The _typed_ kind, which we intend to discuss in a follow-up article, is a restricted version.
+It doesn't have the logical paradox that plagues the untyped kind and can serve as a logic system.
+
+The list of languages based on lambda calculus includes:
+
+- Haskell
+- Most languages in the ML family, like OCaml and F#
+- Cardano blockchain's Plutus Core
+
```

Comment:
I would add dependently typed languages as well

Reply:
It's complicated with Lisp. The "wisdom of the crowds" insists McCarthy based Lisp on λ-calculus, but the man himself insists this was not the case. I'm inclined to believe the latter, because while it's obvious McCarthy was _aware_ of λ-calculus, AFAIK the first versions of Lisp didn't really fit into any λ-calculus evaluation strategy. Modern Lisp dialects are arguably similar enough to λ-calculus to say they're loosely based on it, but historically it seems to be more of a case of convergent evolution. It is pretty hard to untangle what inspired what, though.

Reply:
As for "lambdas as anonymous functions", it's also kind of complicated. While the _syntax_ for anonymous functions is often loosely inspired by λ-calculus (often very indirectly), the actual semantics have little to do with it, so I don't know how useful it is to mention that.